src/main/java/net/neoforged/neoforge/common/extensions/IItemExtension.java

```diff
@@ -422,17 +425,43 @@ default int getEnchantmentValue(ItemStack stack) {
      * either from the enchantment table or other random enchantment mechanisms.
      * As a special case, books are primary items for every enchantment.
      * <p>
-     * Other application mechanisms, such as the anvil, check {@link Enchantment#isSupportedItem(ItemStack)} instead.
-     * If you want those mechanisms to be able to apply an enchantment, you will need to add your item to the relevant tag.
+     * Other application mechanisms, such as the anvil, check {@link #isSupportedItemFor(ItemStack, Holder)} instead.
+     * If you want those mechanisms to be able to apply an enchantment, you will need to add your item to the relevant tag or override that method.
      *
      * @param stack       the item stack to be enchanted
      * @param enchantment the enchantment to be applied
      * @return true if this item should be treated as a primary item for the enchantment
      * @apiNote Call via {@link IItemStackExtension#isPrimaryItemFor(Holder)}
+     * 
+     * @see #isSupportedItemFor(ItemStack, Holder)
      */
     @ApiStatus.OverrideOnly
     default boolean isPrimaryItemFor(ItemStack stack, Holder<Enchantment> enchantment) {
-        return stack.getItem() == Items.BOOK || enchantment.value().isPrimaryItem(stack);
+        if (stack.getItem() == Items.BOOK) {
+            return true;
+        }
+        Optional<HolderSet<Item>> primaryItems = enchantment.value().definition().primaryItems();
+        return this.isSupportedItemFor(stack, enchantment) && (primaryItems.isEmpty() || stack.is(primaryItems.get()));
```

Comment:
Is primaryItems.isEmpty() needed here? An empty holderSet should cause the contains to return false i thought

Reply:
`primaryItems` is an Optional - if not present, any supported item is accepted as primary